### PR TITLE
feat: Deploy HTML docs to the docs branch of Bitbucket

### PR DESCRIPTION
**Pull request by SwathiSrirangapatnaAshokKumar-TomTom**
Deploy the Sphinx built HTML files to the docs branch of Bitbucket using https://github.com/davisp/ghp-import.
**ghp-import** will remove all the files located in the remote "docs" branch and then push the generated HTML files to the docs branch.

### CODE REVIEW

File: vars/pythonSetupPyPipeline.groovy [hunk 1]
```diff
@@ -215,6 +215,26 @@ def call(Map pipelineParams) {
...
+              beforeAgent true
+              allOf {
+                branch "master"
+                expression { params.doRelease }
+              }
+            }
+            steps {
+              sshagent([pipelineParams.sshAgentUser]) {
+                script {
+                  sh "ghp-import -m \"Documentation update to $moduleVersion\" -p -b docs build/sphinx/html"
```
Comment (maikelvdh) [inline vars/pythonSetupPyPipeline.groovy:243]: You will also be adding this to https://github.com/tomtom-international/cookiecutter-python/blob/master/%7B%7Bcookiecutter.project_name%7D%7D/requirements_dev.txt I suppose, so that it will become part of the requirements for development?

File: vars/pythonSetupPyPipeline.groovy [hunk 2]
```diff
@@ -215,6 +215,26 @@ def call(Map pipelineParams) {
                                 pipelineParams.dockerRepo, moduleName, moduleVersion)
             }
           } // Deploy Docker
+
+          // Deploy the html documents to the docs branch, which when using "BitBucket Pages" allows serving the documentation directly
```
Comment (maikelvdh) [inline vars/pythonSetupPyPipeline.groovy:219]: @martinm82 There needs to be an assessment of the impact of this change for other python repositories (both GitHub and BitBucket based). Can you help here with?

File: vars/pythonSetupPyPipeline.groovy (hunk 1 above)
Comment (SwathiSrirangapatnaAshokKumar-TomTom) [reply vars/pythonSetupPyPipeline.groovy:243]: yes, I need to add **ghp-import** to the https://github.com/tomtom-international/cookiecutter-python/blob/master/%7B%7Bcookiecutter.project_name%7D%7D/requirements_dev.txt.
Also to other repos generated from **cookiecutter-python** template.

File: vars/pythonSetupPyPipeline.groovy (hunk 2 above)
Comment (martinm82) [reply vars/pythonSetupPyPipeline.groovy:219]: We could in fact check here whether `ghp-import` is available, if not skip this stage and dump a warning that no docs will be generated due to a repo that is not aligned with the cookiecutter-python template anymore and that the user should re-run against it.

**Review by martinm82 — COMMENTED** [review]
Do you have an example how this looks like in Github? I saw already the example in BitBucket server which is super cool 👍

File: vars/pythonSetupPyPipeline.groovy (hunk 2 above)
Comment (SwathiSrirangapatnaAshokKumar-TomTom) [reply vars/pythonSetupPyPipeline.groovy:219]: I added a condition `isDeployDocsPluginInstalled()` (line 228) to check if `ghp-import` is available and skip the stage.

**Comment by SwathiSrirangapatnaAshokKumar-TomTom** [comment]
> Do you have an example how this looks like in Github? I saw already the example in BitBucket server which is super cool +1

I tried generating [GitHub Pages](https://swathisa.github.io/ebr-connector/) for https://github.com/tomtom-international/ebr-connector but there is some issue with the rendering.

https://aev.github.io/CoRTEX_ADTF2Generator/ is a good example on how it looks on Github but not sure if it is published using `ghp-import`.

File: vars/pythonSetupPyPipeline.groovy [hunk 3]
```diff
@@ -337,3 +361,12 @@ def deployDockerImage(dockerRegistryUrl, dockerRegistryCredentialsId, dockerFile
     }
   }
 }
+
+def isDeployDocsPluginInstalled() {
+  if (sh(script: "pip list | grep -F ghp-import", returnStatus: true) == 0) {
```
Comment (martinm82) [inline vars/pythonSetupPyPipeline.groovy]: If you would use `pip show ghp-import` you would not need to use grep and instead rely directly on the exit code (0 if available or 1 if not).

I think this should work as well:

```
return sh(script: "pip show ghp-import", returnStatus: true) == 0
```

**Review by martinm82 — CHANGES_REQUESTED** [review]

File: vars/pythonSetupPyPipeline.groovy (hunk 3 above)
Comment (SwathiSrirangapatnaAshokKumar-TomTom) [reply vars/pythonSetupPyPipeline.groovy]: @martinm82 , I applied your suggestion.

**Review by martinm82 — COMMENTED** [review]
Please validate the changes in one of our internal Python projects

File: vars/pythonSetupPyPipeline.groovy [hunk 4]
```diff
@@ -218,8 +218,14 @@ def call(Map pipelineParams) {
...
           stage("Deploy Docs") {
+            agent {
+              docker {
+                image buildImageName
+                args pipelineParams.dockerRunArgs
+                reuseNode true
+              }
+            }
             when {
-              beforeAgent true
```
Comment (martinm82) [inline vars/pythonSetupPyPipeline.groovy]: Why did you remove this? This ensures that the below conditions are evaluated first before the agent is actually used.

See https://jenkins.io/doc/book/pipeline/syntax/

> If beforeAgent is set to true, the when condition will be evaluated first, and the agent will only be entered if the when condition evaluates to true.

Comment (SwathiSrirangapatnaAshokKumar-TomTom) [reply vars/pythonSetupPyPipeline.groovy]: To check if ghp-import plugin is installed in the when condition, the agent has to be used. So, had to remove this beforeAgent being set to true.

Comment (martinm82) [reply vars/pythonSetupPyPipeline.groovy]: Can you please add this explanation to the code.

**Comment by SwathiSrirangapatnaAshokKumar-TomTom** [comment]
> Please validate the changes in one of our internal Python projects

done

**Review by martinm82 — DISMISSED** [review]

**Review by maikelvdh — DISMISSED** [review]

**Review by martinm82 — APPROVED** [review]
